<a href="https://colab.research.google.com/github/Recepurkun/GMKA_Recep_Urkun/blob/master/Odev3_Python_Veri_Analizi_Egitimi__Recep_%C3%9Crk%C3%BCn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center> Güney Marmara ve TRAI
# <center> Python Veri Analizi Eğitimi
# <center> Ödev 3

In [250]:
import yfinance as yf
import plotly.express as px
import plotly.graph_objects as go
import pandas as pd
import matplotlib.pyplot as plt
from plotly.tools import make_subplots

import warnings
warnings.filterwarnings("ignore")

### 2021 EUR/TRY kur tablosunun incelenmesi

1. <i> Yahoo Finance</i> sitesinden 2021 yılına ait EUR/TRY kur değerlerini indiriniz

In [281]:
#EUR/TRY Kur veri cekimi
Kur= yf.download("EURTRY=X", start= "2021-01-01", end = "2022-01-01")
Kur

[*********************100%***********************]  1 of 1 completed


Open      High       Low     Close  Adj Close  Volume
Date                                                                 
2021-01-01   9.08470   9.08470   9.08470   9.08470    9.08470       0
2021-01-04   9.11047   9.11154   8.99617   9.10568    9.10568       0
2021-01-05   9.09273   9.11239   9.04298   9.08882    9.08882       0
2021-01-06   9.08044   9.09500   8.92864   9.08177    9.08177       0
2021-01-07   9.00475   9.07176   8.88154   9.00614    9.00614       0
...              ...       ...       ...       ...        ...     ...
2021-12-27  12.28318  13.03196  12.16917  12.23056   12.23056       0
2021-12-28  12.96156  13.47629  12.88283  13.05254   13.05254       0
2021-12-29  13.35232  14.37038  13.31566  13.35252   13.35252       0
2021-12-30  14.32090  15.11444  14.05750  14.31055   14.31055       0
2021-12-31  14.99399  15.40568  14.69452  14.99306   14.99306       0

[261 rows x 6 columns]

2. Veriyi özet metodları ile inceleyiniz

In [265]:
Kur.head(5)

Open     High      Low    Close  Adj Close  Volume
Date                                                             
2021-01-01  9.08470  9.08470  9.08470  9.08470    9.08470       0
2021-01-04  9.11047  9.11154  8.99617  9.10568    9.10568       0
2021-01-05  9.09273  9.11239  9.04298  9.08882    9.08882       0
2021-01-06  9.08044  9.09500  8.92864  9.08177    9.08177       0
2021-01-07  9.00475  9.07176  8.88154  9.00614    9.00614       0

In [266]:
Kur.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 261 entries, 2021-01-01 to 2021-12-31
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       261 non-null    float64
 1   High       261 non-null    float64
 2   Low        261 non-null    float64
 3   Close      261 non-null    float64
 4   Adj Close  261 non-null    float64
 5   Volume     261 non-null    int64  
dtypes: float64(5), int64(1)
memory usage: 14.3 KB


In [267]:
Kur.describe()

Open        High         Low       Close   Adj Close  Volume
count  261.000000  261.000000  261.000000  261.000000  261.000000   261.0
mean    10.452091   10.592024   10.363882   10.456588   10.456588     0.0
std      1.773753    1.959326    1.706138    1.772268    1.772268     0.0
min      8.386340    8.464750    8.354830    8.387260    8.387260     0.0
25%      9.627840    9.695800    9.553790    9.629450    9.629450     0.0
50%     10.090770   10.149140   10.048100   10.092100   10.092100     0.0
75%     10.441610   10.472200   10.359720   10.445100   10.445100     0.0
max     18.671040   20.733000   18.657669   18.655109   18.655109     0.0

3. Veri temizleme: <br>
	• Tabloda sadece "Adj Close" (düzeltilmiş kapanış) sütununu bırakınız.<br>
	• "Adj Close" sütununu " Kapanış (TRY)" olarak yeniden adlandırınız.

In [282]:
Kur = Kur[['Adj Close']] #sadece istedigimiz sutunu gorebiliyoruz
Kur.columns = ["EUR/TRY"]
Kur

EUR/TRY
Date                
2021-01-01   9.08470
2021-01-04   9.10568
2021-01-05   9.08882
2021-01-06   9.08177
2021-01-07   9.00614
...              ...
2021-12-27  12.23056
2021-12-28  13.05254
2021-12-29  13.35252
2021-12-30  14.31055
2021-12-31  14.99306

[261 rows x 1 columns]

4. Tabloya yeni sütun ekleme:<br>
	• Tabloya yeni bir sütun ekleyerek Kapanış (TRY) değerlerini 1 Ocak 2021 tarihindeki kapanış değerine göre normalize ediniz.<br>
	• BIST30 endeksinin 2021 yılı boyunca değişimini yorumlayınız.

In [283]:
Kur['Normalize EUR/TRY'] = Kur["EUR/TRY"]/Kur.iloc[0,0] #ustteki yapilan islem ile tum degerleri norm'a gore boluyoruz boylece butun veriler pesi sira diziliyor.
Kur

EUR/TRY  Normalize EUR/TRY
Date                                   
2021-01-01   9.08470           1.000000
2021-01-04   9.10568           1.002309
2021-01-05   9.08882           1.000454
2021-01-06   9.08177           0.999678
2021-01-07   9.00614           0.991353
...              ...                ...
2021-12-27  12.23056           1.346281
2021-12-28  13.05254           1.436761
2021-12-29  13.35252           1.469781
2021-12-30  14.31055           1.575236
2021-12-31  14.99306           1.650364

[261 rows x 2 columns]

In [355]:
#BIST30 verilerini indirip duzenledik
B30 = yf.download("XU030.IS", start="2021-01-01", end="2021-12-31")
B30 = B30[["Adj Close"]]
B30.columns = ["Kapanis (TRY)"]
B30["Normalize"] = B30["Kapanis (TRY)"]/ B30.iloc[0,0]
B30

[*********************100%***********************]  1 of 1 completed


Kapanis (TRY)  Normalize
Date                                
2021-01-04    1652.900024   1.000000
2021-01-05    1640.400024   0.992438
2021-01-06    1657.699951   1.002904
2021-01-07    1678.199951   1.015306
2021-01-08    1694.400024   1.025107
...                   ...        ...
2021-12-24    2086.800049   1.262508
2021-12-27    2089.399902   1.264081
2021-12-28    2032.900024   1.229899
2021-12-29    2086.300049   1.262206
2021-12-30    2040.199951   1.234315

[238 rows x 2 columns]

In [356]:
#BIST30'un Normalize grafigi
fig_1 = px.line(B30, x=B30.index, y=B30.Normalize,
        width=1000, height=500,
        title = "BIST30 Endeksi 2021 - Normalize (1 ocak 2021)",
        hover_data = {"Normalize": ":.2f", "Kapanis (TRY)": ":.2f",},
        labels = {"Date" : "Tarih", "Normalize" : "BIST 30 Endeksi (TRY)"}
        )
fig_1.show()

# ***BIST 30 Endeksi 2021 yili sonlarına kadar standart inis cikis egilimleri gostermesine karsin yil sonuna dogru ivme kazanarak en yuksek 1.53 katina cikmistir. Seneyi ise 1.23 kat yukselerek kapatmistir.***


5. Veriyi grafik olarak inceleyiniz:<br>
	• Tabloyu px grafiği olarak görselleştiriniz.<br>
	• Tabloyu go. grafiği olarak görselleştiriniz.

In [357]:
#BIST30 ve Kur verilerini ayni tabloda gorebilmek icin birlestiriyoruz.
B30K_2021 = B30.merge(Kur, left_index= True, right_index=True)
B30K_2021
#Normalize EUR/TRY'nin 1den degil de 1.002309 olarak gozukme sebebi Borsanin 1.ayin 4unde acilmis olmasi.

Kapanis (TRY)  Normalize   EUR/TRY  Normalize EUR/TRY
Date                                                             
2021-01-04    1652.900024   1.000000   9.10568           1.002309
2021-01-05    1640.400024   0.992438   9.08882           1.000454
2021-01-06    1657.699951   1.002904   9.08177           0.999678
2021-01-07    1678.199951   1.015306   9.00614           0.991353
2021-01-08    1694.400024   1.025107   8.99850           0.990512
...                   ...        ...       ...                ...
2021-12-24    2086.800049   1.262508  12.79752           1.408689
2021-12-27    2089.399902   1.264081  12.23056           1.346281
2021-12-28    2032.900024   1.229899  13.05254           1.436761
2021-12-29    2086.300049   1.262206  13.35252           1.469781
2021-12-30    2040.199951   1.234315  14.31055           1.575236

[238 rows x 4 columns]

In [358]:
#B30k'nin px grafigi
px.line(B30K_2021, y=["Normalize","Normalize EUR/TRY"])

In [359]:
#B30K'nin go. grafigi
fig = go.Figure()#oncelikle figur olusturuyoruz
fig.add_trace(go.Bar(x=B30K_2021.index, y= B30K_2021["Normalize"], name="BIST30")) #katman eklemek, bar grafigi cizdiriyoruz
fig.add_trace(go.Scatter(x=B30K_2021.index, y= B30K_2021["Normalize EUR/TRY"], name="EUR/TRY")) #cizgi grafigi cizdiriyoruz

fig.update_layout(title="Normalize BIST 100 ve Normalize Kur",
                  xaxis_title = "Tarih", xaxis_dtick="M1", #dtick ile butun aylari gostermesini sagliyoruz
                  yaxis_title = "Kapanis (TRY)")

### 2021 ve 2022 değerlerini birlikte inceleme

1. <i> Yahoo Finance</i> sitesinden 2022 BIST30 ve EUR/TRY kur değerlerini indiriniz

In [360]:
indirme_listesi = ["XU100.IS","EURTRY=X"]
B30K_2022 = pd.DataFrame()

for i in indirme_listesi:
  data = yf.download(i, start = "2022-01-01", end= "2023-01-01")
  data = data["Adj Close"].rename(i)
  B30K_2022 = pd.concat([B30K_2022, data], axis=1) #verilen 2 veriyi sutun olarak birlestirme islemi

B30K_2022

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


XU100.IS   EURTRY=X
2022-01-03  1926.699951  15.189530
2022-01-04  1963.599976  14.763000
2022-01-05  2005.300049  15.071520
2022-01-06  2007.599976  15.445910
2022-01-07  2033.300049  15.434040
...                 ...        ...
2022-05-19          NaN  16.713699
2022-07-11          NaN  17.503019
2022-07-12          NaN  17.325560
2022-07-15          NaN  17.506651
2022-08-30          NaN  18.190140

[260 rows x 2 columns]

2. Indirdiğiniz verileri 2021 verilerini temizleyip düzenlediğiniz üzere hazırlayınız.

In [361]:
B30K_2022.columns = ["Kapanis (TRY)","EUR/TRY"]#ve satir isimlendirmelerini degistiriyoruz
B30K_2022 = B30K_2022[B30K_2022["Kapanis (TRY)"]>0]#borsanin kapali oldugu zaman dilimini tablomuzdan siliyoruz.
B30K_2022

Kapanis (TRY)    EUR/TRY
2022-01-03    1926.699951  15.189530
2022-01-04    1963.599976  14.763000
2022-01-05    2005.300049  15.071520
2022-01-06    2007.599976  15.445910
2022-01-07    2033.300049  15.434040
...                   ...        ...
2022-12-26    5504.100098  19.734200
2022-12-27    5434.500000  19.815269
2022-12-28    5345.600098  19.805389
2022-12-29    5476.899902  19.867519
2022-12-30    5509.200195  19.937571

[252 rows x 2 columns]

3. 2021 ve 2022 veri tablolarını tek bir tablo haline getiriniz.

In [372]:
B30K_2021 = B30K_2021[["Kapanis (TRY)","EUR/TRY"]]
B30K_2021

Kapanis (TRY)   EUR/TRY
Date                               
2021-01-04    1652.900024   9.10568
2021-01-05    1640.400024   9.08882
2021-01-06    1657.699951   9.08177
2021-01-07    1678.199951   9.00614
2021-01-08    1694.400024   8.99850
...                   ...       ...
2021-12-24    2086.800049  12.79752
2021-12-27    2089.399902  12.23056
2021-12-28    2032.900024  13.05254
2021-12-29    2086.300049  13.35252
2021-12-30    2040.199951  14.31055

[238 rows x 2 columns]

In [374]:
Veri_21_22 = pd.concat([B30K_2021, B30K_2022], axis = 0)#2022 ile 2023 yillarinin verilerini birlestiriyoruz
Veri_21_22

Kapanis (TRY)    EUR/TRY
2021-01-04    1652.900024   9.105680
2021-01-05    1640.400024   9.088820
2021-01-06    1657.699951   9.081770
2021-01-07    1678.199951   9.006140
2021-01-08    1694.400024   8.998500
...                   ...        ...
2022-12-26    5504.100098  19.734200
2022-12-27    5434.500000  19.815269
2022-12-28    5345.600098  19.805389
2022-12-29    5476.899902  19.867519
2022-12-30    5509.200195  19.937571

[490 rows x 2 columns]

In [376]:
Veri_21_22["Normalize"] = Veri_21_22["Kapanis (TRY)"]/Veri_21_22.iloc[0,0]
Veri_21_22["Normalize EUR/TRY"] = Veri_21_22["EUR/TRY"]/Veri_21_22.loc[Veri_21_22.index[0],"EUR/TRY"]#once verinin indexini sec sonra da usd/try sutununu kullanmak istiyoruz diyoruz
Veri_21_22["degisim"] = Veri_21_22["Normalize"].diff(1)
Veri_21_22["d_yon"] = Veri_21_22["degisim"].map(lambda x: 'dusen gun' if x < 0 else 'yukselen gun') 
Veri_21_22["degisim%"] = (Veri_21_22["degisim"]/Veri_21_22["Normalize"].shift(1))*100
Veri_21_22

Kapanis (TRY)    EUR/TRY  Normalize  Normalize EUR/TRY   degisim  \
2021-01-04    1652.900024   9.105680   1.000000           1.000000       NaN   
2021-01-05    1640.400024   9.088820   0.992438           0.998148 -0.007562   
2021-01-06    1657.699951   9.081770   1.002904           0.997374  0.010466   
2021-01-07    1678.199951   9.006140   1.015306           0.989068  0.012402   
2021-01-08    1694.400024   8.998500   1.025107           0.988229  0.009801   
...                   ...        ...        ...                ...       ...   
2022-12-26    5504.100098  19.734200   3.329966           2.167241  0.029827   
2022-12-27    5434.500000  19.815269   3.287858           2.176144 -0.042108   
2022-12-28    5345.600098  19.805389   3.234073           2.175059 -0.053784   
2022-12-29    5476.899902  19.867519   3.313509           2.181882  0.079436   
2022-12-30    5509.200195  19.937571   3.333051           2.189575  0.019542   

                   d_yon  degisim%  
2021-01-04  yukselen gun       NaN  
2021-01-05     dusen gun -0.756247  
2021-01-06  yukselen gun  1.054616  
2021-01-07  yukselen gun  1.236653  
2021-01-08  yukselen gun  0.965324  
...                  ...       ...  
2022-12-26  yukselen gun  0.903797  
2022-12-27     dusen gun -1.264514  
2022-12-28     dusen gun -1.635843  
2022-12-29  yukselen gun  2.456222  
2022-12-30  yukselen gun  0.589755  

[490 rows x 7 columns]

4. Veriyi grafik olarak inceleyiniz:<br>
	• Tabloları px grafiği olarak görselleştiriniz.<br>
	• Tabloları go. grafiği olarak görselleştiriniz.<br>
	• Tabloları subplots grafiği olarak görselleştiriniz.

In [391]:
#PX GRAFIGI - B30 Normalize
fig_4 = px.line(Veri_21_22, y="Normalize",
        width = 1000, height=500,
        title="2021-2022 | BIST30 Gelişimi",
        hover_data={"Normalize" : ':.2f',"Kapanis (TRY)" : ":.0f"},
        labels = {"Normalize" : "Normalize (1 Ocak 2021 = 1)","index" : "Tarih"})
fig_4.show()

In [393]:
#PX GRAFIGI - Normalize EUR/TRY
fig_5 = px.line(Veri_21_22, y="Normalize EUR/TRY",
        width = 1000, height=500,
        title="2021-2022 | EUR/TRY Gelişimi",
        hover_data={"EUR/TRY" : ':.2f',"Kapanis (TRY)" : ":.0f"},
        labels = {"Normalize" : "Normalize (1 Ocak 2021 = 1)","index" : "Tarih"})
fig_5.show()

In [396]:
#px 2021-2022 | BIST30 ve EUR/TRY Gelişimi
px.line(Veri_21_22, y=["Normalize","Normalize EUR/TRY"])

In [397]:
#GO. GRAFIGI
fig = go.Figure()
fig.add_trace(go.Bar(x=Veri_21_22.index, y= Veri_21_22["Normalize"], name="BIST30")) #katman eklemek, bar grafigi cizdiriyoruz
fig.add_trace(go.Scatter(x=Veri_21_22.index, y= Veri_21_22["Normalize EUR/TRY"], name="EUR/TRY")) #cizgi grafigi cizdiriyoruz

fig.update_layout(title="2021-20211 BIST30 ve EUR/TRY Gelişimi",
                  xaxis_title = "Tarih", xaxis_dtick="M1",
                  yaxis_title = "Normalize (1 Ocak 2021 = 1)")

In [398]:
#subplots GRAFIGI
fig = make_subplots(rows=1, cols=2)
fig.add_trace(go.Scatter(x = Veri_21_22.index , y= Veri_21_22["Normalize"], name="Normalize"),
              row=1, col=1)
fig.add_trace(go.Bar(x = Veri_21_22.index , y= Veri_21_22["Normalize EUR/TRY"], name="Normalize EUR/TRY"),
              row=1, col=2)
fig.show()